File: lib/test/t_PolygonMesher_std.cxx
#include <iostream>
#include <cmath>

#include "otmeshing/otmeshing.hxx"
#include "otmeshing/PolygonMesher.hxx"

using namespace OT;
using namespace OTMESHING;

int main()
{
  PolygonMesher mesher;
  std::cout << "mesher=" << mesher << std::endl;

  // 2D convex polygon: unit square
  Sample points(4, 2);
  points(0, 0) = 0.0; points(0, 1) = 0.0;
  points(1, 0) = 1.0; points(1, 1) = 0.0;
  points(2, 0) = 1.0; points(2, 1) = 1.0;
  points(3, 0) = 0.0; points(3, 1) = 1.0;

  Mesh mesh(mesher.build(points));
  std::cout << "mesh=" << mesh << std::endl;
  if (mesh.getDimension() != 2)
    throw std::runtime_error("Expected dimension 2");
  if (mesh.getSimplices().getSize() != 2)
    throw std::runtime_error("Expected 2 simplices");
  if (!mesh.isValid())
    throw std::runtime_error("Expected valid mesh");
  if (std::abs(mesh.getVolume() - 1.0) > 1e-14)
    throw std::runtime_error("Expected volume 1.0");

  // 3D coplanar points (z=0)
  Sample points3d(4, 3);
  points3d(0, 0) = 0.0; points3d(0, 1) = 0.0; points3d(0, 2) = 0.0;
  points3d(1, 0) = 1.0; points3d(1, 1) = 0.0; points3d(1, 2) = 0.0;
  points3d(2, 0) = 1.0; points3d(2, 1) = 1.0; points3d(2, 2) = 0.0;
  points3d(3, 0) = 0.0; points3d(3, 1) = 1.0; points3d(3, 2) = 0.0;

  Mesh mesh3d(mesher.build(points3d));
  std::cout << "mesh3d=" << mesh3d << std::endl;
  if (mesh3d.getDimension() != 3)
    throw std::runtime_error("Expected dimension 3");
  if (mesh3d.getSimplices().getSize() != 2)
    throw std::runtime_error("Expected 2 simplices");
  if (!mesh3d.isValid())
    throw std::runtime_error("Expected valid mesh");

  // triangle (minimal case)
  Sample triangle(3, 2);
  triangle(0, 0) = 0.0; triangle(0, 1) = 0.0;
  triangle(1, 0) = 1.0; triangle(1, 1) = 0.0;
  triangle(2, 0) = 0.0; triangle(2, 1) = 1.0;

  Mesh triMesh(mesher.build(triangle));
  std::cout << "triMesh=" << triMesh << std::endl;
  if (triMesh.getSimplices().getSize() != 1)
    throw std::runtime_error("Expected 1 simplex");
  if (!triMesh.isValid())
    throw std::runtime_error("Expected valid mesh");

  std::cout << "ok" << std::endl;
  return 0;
}

File: python/test/t_PolygonMesher_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott
import otmeshing
import math

ot.TESTPREAMBLE()

mesher = otmeshing.PolygonMesher()
print("mesher=", mesher)

# 2d triangulation of a convex polygon (regular n-sided)
polyline = []
n = 20
for i in range(n):
    r = 1.0
    theta = i * 2.0 * math.pi / n
    x = r * math.cos(theta)
    y = r * math.sin(theta)
    polyline.append([x, y])
triangulation = mesher.build(polyline)
print("triangulation=", repr(triangulation))
assert triangulation.getDimension() == 2
assert len(triangulation.getSimplices()) == n - 2
assert triangulation.isValid()
# area of unit regular n-gon: n/2 * sin(2*pi/n)
area_ref = 0.5 * n * math.sin(2.0 * math.pi / n)
ott.assert_almost_equal(triangulation.getVolume(), area_ref)

# 2d triangulation of a non-convex polygon (snail-like)
polyline = [
    [0, 0],
    [0, 5],
    [6, 5],
    [6, 0],
    [2, 0],
    [2, 3],
    [4, 3],
    [4, 2],
    [3, 2],
    [3, 1],
    [5, 1],
    [5, 4],
    [1, 4],
    [1, 0],
]
triangulation = mesher.build(polyline)
print("triangulation=", repr(triangulation))
assert triangulation.getDimension() == 2
assert len(triangulation.getSimplices()) == 12
assert triangulation.isValid()
# area of this polygon: outer box 6x5 minus inner cutout 4x4 + some
# compute numerically: polygon area using shoelace formula
xs = [p[0] for p in polyline]
ys = [p[1] for p in polyline]
area_ref = 0.5 * abs(
    sum(
        xs[i] * ys[(i + 1) % len(polyline)] - xs[(i + 1) % len(polyline)] * ys[i]
        for i in range(len(polyline))
    )
)
ott.assert_almost_equal(triangulation.getVolume(), area_ref)

# simple triangle (minimal case)
polyline = [[0.0, 0.0], [1.0, 0.0], [0.0, 1.0]]
triangulation = mesher.build(polyline)
print("triangulation=", repr(triangulation))
assert triangulation.getDimension() == 2
assert len(triangulation.getSimplices()) == 1
assert triangulation.isValid()
ott.assert_almost_equal(triangulation.getVolume(), 0.5)

# 2D polygon with points in clockwise order
polyline = [[0.0, 0.0], [0.0, 1.0], [1.0, 1.0], [1.0, 0.0]]
triangulation = mesher.build(polyline)
print("triangulation=", repr(triangulation))
assert triangulation.getDimension() == 2
assert len(triangulation.getSimplices()) == 2
assert triangulation.isValid()
ott.assert_almost_equal(triangulation.getVolume(), 1.0)

# 3D coplanar points (z=0 plane)
polyline = [[0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [1.0, 1.0, 0.0], [0.0, 1.0, 0.0]]
triangulation = mesher.build(polyline)
print("triangulation=", repr(triangulation))
assert triangulation.getDimension() == 3
assert len(triangulation.getSimplices()) == 2
assert triangulation.isValid()
ott.assert_almost_equal(triangulation.getVolume(), 0.0)

# check that the mesh simplices have the right size (dimension+1)
for simplex in triangulation.getSimplices():
    assert len(simplex) == 4  # 3D -> 4 vertices per simplex

# 5D coplanar points
polyline = [
    [0.0, 0.0, 0.0, 0.0, 0.0],
    [1.0, 0.0, 0.0, 0.0, 0.0],
    [1.0, 1.0, 0.0, 0.0, 0.0],
    [0.0, 1.0, 0.0, 0.0, 0.0],
]
triangulation = mesher.build(polyline)
print("triangulation=", repr(triangulation))
assert triangulation.getDimension() == 5
assert len(triangulation.getSimplices()) == 2
assert triangulation.isValid()

# error cases
# too few points
try:
    mesher.build([[0.0, 0.0], [1.0, 0.0]])
    assert False
except RuntimeError:
    pass

# redundant vertex
try:
    mesher.build([[0.0, 0.0], [1.0, 0.0], [0.0, 1.0], [1.0, 0.0]])
    assert False
except RuntimeError:
    pass

# collinear points (intrinsic dimension 1)
try:
    mesher.build([[0.0, 0.0], [1.0, 0.0], [2.0, 0.0]])
    assert False
except RuntimeError:
    pass
